### PR TITLE
json output: fix number format

### PR DESCRIPTION
**Pull request by vchlum**
#### Describe Bug or Feature

While constructing JSON output, function strtod() is not enough to check number is suitable for JSON output. The current additional check only checks the number is not containing a leading zero.  Incorrect formats are also:  .123,123.,00,-00,00.123,-00.123, and possibly others. See the problem:
```
vchlum@openpbs-ptl:~> echo "sleep 1000" | qsub -v FOO1=.123,FOO2=123.,FOO3=00,FOO4=-00,FOO5=00.123,FOO6=-00.123
69.openpbs-ptl
vchlum@openpbs-ptl:~> qstat -f -F json | grep FOO | grep -v Submit_arguments
                "FOO1":.123,
                "FOO2":123.,
                "FOO3":00,
                "FOO4":-00,
                "FOO5":00.123,
                "FOO6":-00.123,
vchlum@openpbs-ptl:~> python3 -c "import json;json.loads(\"$(qstat -f -F json | tr -d "\n" | sed 's/["]/\\&/g')\")"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1853 (char 1852)
vchlum@openpbs-ptl:~> 
```

PTL test TestQstatFormats.test_qstat_json_valid_job_special_env is currently not working. **It is false positive.** Should be fixed in #2602.

#### Describe Your Change

In this patch, we enumerate number formats that are considered to be a number in JSON, string otherwise. The allowed number formats are: `0`, `-0`, `0\.[0-9]+`, `-0\.[0-9]+`, `[1-9]+[0-9]*\.?[0-9]*`, and `-[1-9]+[0-9]*\.?[0-9]*`

#### Attach Test and Valgrind Logs/Output

Improved ptl test (after applying #2602): [ptl_output.txt](https://github.com/openpbs/openpbs/files/11893530/ptl_output.txt)

Manual test:
```
vchlum@openpbs-ptl:~> echo "sleep 1000" | qsub -v FOO1=.123,FOO2=123.,FOO3=00,FOO4=-00,FOO5=00.123,FOO6=-00.123,FOO7=0,FOO8=-0,FOO9=0.000,FOO10=1 
74.openpbs-ptl
vchlum@openpbs-ptl:~> qstat -f -F json | grep FOO | grep -v Submit_arguments
                "FOO1":".123",
                "FOO2":"123.",
                "FOO3":"00",
                "FOO4":"-00",
                "FOO5":"00.123",
                "FOO6":"-00.123",
                "FOO7":0,
                "FOO8":-0,
                "FOO9":0.000,
                "FOO10":1,
vchlum@openpbs-ptl:~> python3 -c "import json;json.loads(\"$(qstat -f -F json | tr -d "\n" | sed 's/["]/\\&/g')\")"
vchlum@openpbs-ptl:~>
```

### CODE REVIEW

**Review by nithinj — APPROVED** [review]
LGTM!
we should make use of cjson for all these.

**Comment by nithinj** [comment]
It is ideal to create a function is_valid_json_number() and expose it using pbs_json.h

**Comment by vchlum** [comment]
OK, I will prepare is_valid_json_number().
BTW: I also agree that using cjson or something would be great. There were several fixes of json in the past and we still can find an error in it. Considering licensing, is cjson ok?

**Comment by nithinj** [comment]
Since CJSON is released under the permissive MIT license, it is generally compatible with the GPL. 
Tagging @crjayadev

**Comment by crjayadev** [comment]
> Since CJSON is released under the permissive MIT license, it is generally compatible with the GPL. Tagging @crjayadev

On a quick look, I don't see an issue with licensing, however I will get this checked. Meanwhile please continue the technical review by maintainers for the use of cjson. I will get back here within a day or two.

**Comment by vchlum** [comment]
[ptl_output.txt](https://github.com/openpbs/openpbs/files/12066409/ptl_output.txt) test of the function is_valid_json_number()

File: src/lib/Libcmds/pbs_json.c [hunk 1]
```diff
@@ -543,3 +518,48 @@ generate_json(FILE *stream)
...
+			(
+				(val >= 1 && val_len > 0 && value[0] != '0') || /* [1-9]+[0-9]*\.?[0-9]* is number */
+				(val <= -1 && val_len > 1 && value[1] != '0') /* -[1-9]+[0-9]*\.?[0-9]* is number */
+			)
+		)) {
+		return 1;
+	}
+
+	return 0;
+}
```
Comment (nithinj) [inline src/lib/Libcmds/pbs_json.c:565]: insert a new line

**Comment by crjayadev** [comment]
> > Since CJSON is released under the permissive MIT license, it is generally compatible with the GPL. Tagging @crjayadev
> 
> On a quick look, I don't see an issue with licensing, however I will get this checked. Meanwhile please continue the technical review by maintainers for the use of cjson. I will get back here within a day or two.

license-wise we are good. cjson can be used.

File: src/lib/Libcmds/pbs_json.c (hunk 1 above)
Comment (vchlum) [reply src/lib/Libcmds/pbs_json.c]: done

**Review by nithinj — APPROVED** [review]

**Comment by vchlum** [comment]
Thank you for the confirmation. I will look into the cjson. Anyway, I suggest using a different PR for cjson. Should we continue the review of this one or close it?

**Review by hirenvadalia — APPROVED** [review]

**Pull request merged**